### PR TITLE
Implement `--extern-location`

### PR DESCRIPTION
**Pull request by jsgf**
This PR implements `--extern-location` as a followup to #72342 as part of the implementation of #57274. The goal of this PR is to allow rustc, in coordination with the build system, to present a useful diagnostic about how to remove an unnecessary dependency from a dependency specification file (eg Cargo.toml).

EDIT: Updated to current PR state.

The location is specified for each named crate - that is, for a given `--extern foo[=path]` there can also be `--extern-location foo=<location>`. It supports ~~three~~ two styles of location:
~~1. `--extern-location foo=file:<path>:<line>` - a file path and line specification
1. `--extern-location foo=span:<path>:<start>:<end>` - a span specified as a file and start and end byte offsets~~
1. `--extern-location foo=raw:<anything>` - a raw string which is included in the output
1. `--extern-location foo=json:<anything>` - an arbitrary Json structure which is emitted via Json diagnostics in a `tool_metadata` field.

~~1 & 2 are turned into an internal `Span`, so long as the path exists and is readable, and the location is meaningful (within the file, etc). This is used as the `Span` for a fix suggestion which is reported like other fix suggestions.~~

`raw` and `json` are for the case where the location isn't best expressed as a file and location within that file. For example, it could be a rule name and the name of a dependency within that rule. `rustc` makes no attempt to parse the raw string, and simply includes it in the output diagnostic text. `json` is only included in json diagnostics. `raw` is emitted as text and also as a json string in `tool_metadata`.

If no `--extern-location` option is specified then it will emit a default json structure consisting of `{"name": name, "path": path}` corresponding to the name and path in `--extern name=path`.

This is a prototype/RFC to make some of the earlier conversations more concrete. It doesn't stand on its own - it's only useful if implemented by Cargo and other build systems. There's also a ton of implementation details which I'd appreciate a second eye on as well.

~~**NOTE** The first commit in this PR is #72342 and should be ignored for the purposes of review. The first commit is a very simplistic implementation which is basically raw-only, presented as a MVP. The second implements the full thing, and subsequent commits are incremental fixes.~~

cc @ehuss @est31 @petrochenkov @estebank

### CODE REVIEW

**Comment by tshepang** [comment]
When I saw `-loc`, my head said "lines of code". Would it be ok to rename to `--extern-location`?

**Comment by jsgf** [comment]
@tshepang Yes, I have a tendency to over-abbr.

**Comment by petrochenkov** [comment]
This is great, thank you.
Diagnostic spans pointing to build system files in particular.
I'll nominate this for the compiler and cargo teams because this needs feedback from some wider audience.

I'll leave a few notes and questions in comments below.

**Comment by petrochenkov** [comment]
### Note 1 - Motivation

This is useful not only for the `unused_crate_dependencies` lint, but for any kind of diagnostics that want to show the point where a crate name was introduced.

All "cannot find or load crate" errors, name ambiguity errors during resolution, lint for private dependencies (https://github.com/rust-lang/rust/issues/44663), perhaps something else.

**Comment by petrochenkov** [comment]
### Note 2 - Dependency tracking

When file is loaded into source map during conversion into a span that file also becomes a dependency for the build.
It's emitted into the depinfo files (`.d`) used by cargo for rebuilds.

Looks like PR puts the files specified in `--extern-location` into source map lazily, only if an error requiring these spans actually happens.
This is great because, AFAIK, cargo doesn't rebuild anything when you touch `Cargo.toml` without introducing some real changes.
If `Cargo.toml` is put into depinfo, then the rebuilds will happen, unless cargo specifically removes it from dependencies.

Lazy loading may be inconvenient in some cases when we want to work with spans uniformly, though.
E.g. you cannot put an item span and a span from `--extern-location` together into a `Vec<Span>` without loading the file and adding it to dependencies.

**Comment by petrochenkov** [comment]
### Note 3 - Relative path

If the path in `--extern-location foo=file:bar.mk` is relative, what should be considered the base directory?

Will cargo and other tools use absolute or relative paths in practice?

**Comment by petrochenkov** [comment]
### Note 4 - Command line length compression

Somewhat related to note 3.

If tools start using absolute paths for `--extern-location`, e.g. `--extern-location foo=span:/my/full/path/Cargo.toml:23:32` rather than `--extern-location foo=span:Cargo.toml:23:32`, then it will be easy to make the command line very long and hit some limits?

In practice all the paths from cargo will refer to the *same* `Cargo.toml`, but different byte ranges inside it.
So, some compression scheme could be useful.
`--extern-location-file my_file=/my/full/path/Cargo.toml --extern-location foo=my_file:23:32 --extern-location bar=my_file:34:43`.

**Comment by petrochenkov** [comment]
### Note 5 - Over-engineering

Are 4 types of extern locations really necessary?
(Leaving this for the teams' feedback.)

**Comment by jsgf** [comment]
@petrochenkov Thank you very much for taking the time to look at this.

1: Yes, I hadn't really considered other uses, but I'm glad there are some.

2: I was concerned about the Spans becoming part of the state of the build, both as dependencies, and for incremental builds. I wasn't sure if it was desirable or not, or what the effects would be. It sounds like it's not an issue, and even desirable.

3: I think the path should be relative to `rustc` cwd, since Cargo (or build system in general) control both the working directory and the paths. They could be absolute, but that would be very redundant. They're also subject to `--remap-path-prefix` which I think is expected.

4: I was a little concerned about command-line expansion. It's proportional to the number of direct dependencies a crate has, which doesn't seem to be large - I haven't done a systematic survey, but my gut says that it's going to be almost always <50. It would be a different matter if it also affected indirect dependencies.

5: I have a local change to remove `raw`, since it's a strict subset of `json`. I'm using `json` for Buck, and I'd expect `cargo` to use `span` if it's going to use it at all, which makes `file` redundant - but it seems like the easiest to use for simpler cases.

One thing I'd like to highlight is that the `json` support includes a schema change for the json diagnostics by adding a `tool_metadata` field. The intent is to allow `json` to be round-tripped through rustc without needing to invent a new encoding/serialization/quoting scheme. It's currently good enough for me to prototype with, but I'd like guidance as to whether there's a better way to handle this (if nothing else I wonder if it should be on just the top-level diagnostic, rather than every child diagnostic).

**Comment by jsgf** [comment]
Thinking about raw vs json a bit more - the `json` type is only emitted in the json format diagnostics and never appears in the human-visible rendered form. But if you want to present something to the user as advice that's not a file coordinate, then it would be useful to have a `text` type for that...

**Comment by est31** [comment]
Why round trip in the first place? Why not use binary depinfo at that point?

**Comment by jsgf** [comment]
@est31

Binary depinfo is very low level - it's just the pathnames of which artifacts it ended up using. At the high level, to auto-fix unused dependencies I need to know which dependencies of which rules need to be removed. So to bridge the gap, I'd need to change buck to:
1. map from pathnames back to the targets which generated them to find the dependencies
2. consume the rustc diagnostics and extend them with the rule/dep info
3. pass those modified diagnostics to the autofix linter

1 isn't possible to do completely because you don't know which alias of a crate it refers to, though I don't think its common to depend on the same crate with multiple names.

2 is pretty awkward because there's currently no mechanism for doing this kind of rewriting, and it would only be possible when using json diagnostics (which is currently not the default).

By comparison using `json` so that buck can pass structured information which is directly usable by the linter though rustc is more correct and much simpler all around, even though it has a bit of extra code in rustc.

**Comment by est31** [comment]
> Binary depinfo is very low level - it's just the pathnames of which artifacts it ended up using.

It would be the level rustc operates on though. For each dependency, you pass a path to rustc via the `--extern` arg. So at the API level of rustc's CLI params, the way to refer to a dependency is through a path.

I agree with your point about using the same file multiple times under different aliases. So depinfo won't be enough indeed to accurately identify a dependency.

What *does* identify a dep though is the tuple of name and binary path, the pair which is passed via the `--extern` arg. What if `--extern-location` is scrubbed entirely and rustc could be made to emit just the extern arg it got in json?

At this point, I believe that this should at least make styles 1 and 2 obsolete. Cargo should implement its own span printing one day anyways for warnings and errors in Cargo.toml.

As for style 3, I get your point about the architecture constraints you are having. I guess it can be kept not because it's best for rustc but because it's best for users like buck. But as styles 1 and 2 aren't really necessary, what about just making raw/json implicit and the only supported mode of `--extern-location`?

**Comment by jsgf** [comment]
@est31 

Yeah, `span` is only there in anticipation of Cargo needing it, it's premature if we're not sure that it does (likewise `file`). I'm fine with leaving them out for the initial patch since we can always add them later if needed.

I don't think `json` should be implicit or the only supported format, though it's certainly the most general. We could have the convention that plain `--extern foo=...` on its own would produce something like `{"crate":"foo","path":"path/to/libfoo.rlib"}`, but using`--extern-location foo=json:{...}` would override this to an arbitrary piece of json.

Since `json` is currently only emitted into json diagnostics, `text` would still be useful as a way of emitting something into the rendered diagnostic.

**Comment by ehuss** [comment]
I just wanted to raise a concern, since T-cargo was tagged on this.  I don't think Cargo will use this anytime soon, so it will be difficult to get a sense of how well it would work in that context. The `unused_crate_dependencies` lint won't fundamentally work with most Cargo projects, so I don't think it is something we'll want to turn on or recommend. @petrochenkov mentioned some other lints that could possibly use it, but I'm uncertain what that would look like or if that would substantially improve those diagnostics.  

Just a minor note about command-line length limits, I think that is a real concern.  I think Windows is probably the most sensitive, with a 32k length limit, with 8k limit on the command-line. A substantial increase means a user would not be able to copy/paste a build command into a terminal. I was investigating this recently for a similar situation (extern URLs for rustdoc), and was considering using `@path` since a relatively small project like Cargo already blows out the 8k limit.

@eddyb It looks like this requires `-Zunstable-options`

**Comment by nikomatsakis** [comment]
Hey @jsgf -- so this PR prompted me to finally get around to writing up the procedure for approving new options to the compiler. You can find it on forge here:

https://forge.rust-lang.org/compiler/new_option.html

The TL;DR is -- can you open an [MCP] with a description of the design? That would in turn create a Zulip topic where we could discuss to decide if it's something we want or if we want to tweak the design.

That said, we're right now [discussing on Zulip](https://zulip-archive.rust-lang.org/131828tcompiler/88059Soyouwanttoaddacompileroption.html) whether to tweak some details of what that says (I'd *love* feedback btw). But all I really want to request is a write-up that says kind of what the option does and why, and maybe covers some of the design discussion -- I think an [MCP] is a decent way to post that but until we've nailed everything down I'm not too particular.

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

**Comment by nikomatsakis** [comment]
@ehuss is there a particular reason cargo wouldn't use this, or is it just "it's not something we were thinking about doing"

**Comment by est31** [comment]
@nikomatsakis --extern-location is mainly useful for the unused deps lint (#72342), but that lint is not useful for cargo, only useful for buck users, simply due to design choices of cargo to have multiple crates (from the perspective of rustc) share the same list of deps.

See https://github.com/rust-lang/rust/pull/72342#issuecomment-631272373

Explaining: If you put some dependency into `[dependencies]` then it might be used by both `lib.rs` as well as `main.rs`, or just by one, or by neither. In the first case, everything is fine, but as rustc gets called separately for main.rs and lib.rs, it's unable to distinguish the second from the third case. In the second case though, there should be no warning but in the third there should.

Same goes for `[dev-dependencies]`. Here you have even more crates from the perspective of rustc that share dependencies (every .rs file in tests/, every .rs file in examples/, etc, the lib.rs file, etc). Only for `build.rs` and `[build-dependencies]` there is no sharing, so theoretically, cargo could pass `-W unused_crate_dependencies` for `build.rs` starting today, however I'd prefer if there would be an unified implementation that works for `[dependencies]`, `[dev-dependencies]` and `[build-dependencies]` in the same way, thus any use of that flag should only be temporary until a unified solution arrives.

Buck on the other hand has a 1:1 matching of dependencies and no such sharing. Thus it can use a rustc native lint. But for cargo, it's not helpful.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/72603#issuecomment-636381491 lists some other motivations besides the `unused_crate_dependencies` lint, which are probably mostly about IDE experience - you select an error or warning related to crates and jump directly into the `Cargo.toml` via span.

>the procedure for approving new options to the compiler.

I'm not sure this is *just* an option.
The buy in from Cargo is pretty important here.
It'll have to add significant amount of information to the command line to get benefits only in case of errors or warnings.

**Comment by est31** [comment]
@petrochenkov IMO a better way would be to not concern rustc with spans to Cargo.toml files but to let rustc print unique identifiers for the crates involved and let cargo figure out which rustc crate was meant, obtain a span for it, and print the span together with a snippet. A rustc command invoked by cargo is already quite long and if each extern had an extern location, it would be even more bloated. It's okay if Buck wants to do it, but cargo shouldn't.

**Comment by nikomatsakis** [comment]
@petrochenkov 

> I'm not sure this is just an option. The buy in from Cargo is pretty important here.

Agreed, this kind of concern is actually precisely why I was asking for people to write-up the motivations for adding the option and create an MCP, so we have a chance to talk it over and make sure we have enough buy-in.

Regarding the points @est31 raised, I see the problem. That's too bad, because I definitely find I accumulate unused dependencies in my crates. I can appreciate the challenge though in usefully linting for them given all the different ways a single bit of source code can be built.

**Comment by est31** [comment]
> That's too bad, because I definitely find I accumulate unused dependencies in my crates.

Cargo should definitely get a native version of this lint some day, it just has different needs than buck. That's why I think the option should be limited to the bare mimimum buck needs (see https://github.com/rust-lang/rust/pull/72603#issuecomment-637895852), as it's only useful for buck and even there it's not needed *technically* but helps to avoid major refactors.

**Comment by jsgf** [comment]
@nikomatsakis Thanks for the pointer. I'm happy to write up an MCP (what does that stand for again?) for this option. I consider this PR very much an exploration of the design space. I have a concrete motivating use-case and environment which I think will give it some real grounding, but it certainly isn't every possible use-case (and definitely not the one that most Rust users will find useful). So I think some discussion will be useful and some refinement is possible.

@est31 It's unfortunate that Cargo's design choices make `-Wunused-crate-dependencies` (and therefore this) not immediately useful. While sharing deps across multiple targets certainly keeps things simple, there's definitely a desire among Cargo users to allow, for example, library and binary dependencies to be split.

**Comment by jsgf** [comment]
@est31 I'm adding an implicit location derived from the --extern name/path tuple, only to find that it actually supports any number of paths:
https://github.com/rust-lang/rust/blob/9c1857fad849ad56d38327b9bc11377a0bdbb4cf/src/librustc_session/config.rs#L374

Does this ever happen in practice? It seems the options are to either grab the first one, or try to list them all. But surely the former is enough for Cargo to work out what's going on (assuming it ever sets more than one path).

**Comment by est31** [comment]
@jsgf 

> I'm adding an implicit location derived from the --extern name/path tuple, only to find that it actually supports any number of paths

Interesting. I've thought a little about this. From the perspective of rustc, the name already uniquely identifies an `--extern` arg (see `parse_externs` function in that file that builds a `BTreeMap<String, ExternEntry>`). So it'd be enough to emit the name of the dep only I think. No paths needed.

For cargo's side, the name wouldn't be an unique identifier, as you could theoretically have dev and build dependencies named the same way in their extern name but pointing to different crates. I'd imagine cargo could provide a mapping from name to unique identifier. That mapping would remain entirely internal to cargo and rustc would only have to complain about dependencies.

> there's definitely a desire among Cargo users to allow, for example, library and binary dependencies to be split.

Indeed, I'm a big supporter of this split as some tools are also published as libraries to crates.io and cause library users to compile deps that only the binary needs. But it would only fix the situation for `[dependencies]`, not for `[dev-dependencies]`.

------

Another issue: currently if you enable the warning, it also causes emission of "warning: 1 warning emitted" like messages. Cargo would either have to intercept them and subtract 1 from the number, or rustc would have to be told to warn about it but not count it as warning...

I'm not really sure what the best design here would be. Any ideas? (question for everyone)

**Comment by jsgf** [comment]
@est31 
> For cargo's side, the name wouldn't be an unique identifier, as you could theoretically have dev and build dependencies named the same way in their extern name but pointing to different crates

Yeah, but that wouldn't happen in a single invocation of `rustc` though, would it? A given compilation is for only one of a library, binary, test or build script.

**Comment by est31** [comment]
> Yeah, but that wouldn't happen in a single invocation of rustc

Indeed. That paragraph was mainly about what cargo has to do. rustc doesn't have to emit any paths at all.

**Comment by jsgf** [comment]
@eddyb It's not a -Z option, but it requires -Zunstable-options to be set to be usable:
https://github.com/rust-lang/rust/blob/3698c08573c60fc1fbf0a3fceb3ad76c77b842c6/src/librustc_session/config.rs#L1649

**Comment by jsgf** [comment]
MCP filed https://github.com/rust-lang/compiler-team/issues/303

**Comment by petrochenkov** [comment]
@jsgf 
>I'm adding an implicit location derived from the --extern name/path tuple, only to find that it actually supports any number of paths ... Does this ever happen in practice?

Yes, it does.
The multiple paths must refer to different versions of the same library, e.g. like this
```
rustc --extern foo=libfoo.rlib --extern foo=libfoo.so --extern foo=libfoo.rmeta
```

File: src/librustc_lint/context.rs
Comment (bjorn3) [inline src/librustc_lint/context.rs:591]: Too much indentation.

**Comment by jsgf** [comment]
@petrochenkov 
>  The multiple paths must refer to different versions of the same library, e.g. like this

When you say "versions" do you mean "different forms of the same library" - ie they're all from the same set of sources and dependencies? If that's the case, then build system perspective they're all equivalent.

**Comment by est31** [comment]
@petrochenkov @jsgf Originally I've thought the path was needed for cargo but I've reconsidered it in my comment [here](https://github.com/rust-lang/rust/pull/72603#issuecomment-640156059) (although it seems I was misunderstood). So please remove the path from the json. Only the name is required. Thanks!

**Comment by petrochenkov** [comment]
@jsgf 
>they're all from the same set of sources and dependencies?

Exactly.

File: src/librustc_lint/context.rs
Comment (jsgf) [reply src/librustc_lint/context.rs:591]: I wonder why rustfmt isn't fixing it.

**Comment by jsgf** [comment]
@est31 I removed the path, but left it as a json object so we can add things later.

**Comment by jsgf** [comment]
Next steps?

I've deployed tooling built around --extern-location within facebook and removed ~2000 unused dependencies across our codebase, and added it to the regular linters so new changes are checked for unused dependencies. So a success story with N=1.

The MCP https://github.com/rust-lang/compiler-team/issues/303 is waiting for a second. My read of this discussion on this PR is generally positive to the idea of this option, but still details to be worked out, I think it's worth a second to second the MCP. @petrochenkov? @est31?

**Comment by nikomatsakis** [comment]
My two cents:

I think it's reasonable to land this even though we've not figured out the cargo integration story, as it provides immediate value, but I'd like to hear from @petrochenkov and others in terms of their satisfaction with the proposal as it stands.

**Comment by nikomatsakis** [comment]
(I would be reluctant to stabilize, however, until we have a clearer picture of what's happening with cargo.)

**Comment by jsgf** [comment]
I'll fold this down to a single commit now that it is locally stable.

Aside from the question of how Cargo might make use of this, I'm also interested in discussing whether this implementation is the best way of using the json diagnostics. I think I'd prefer to have a single field in a top-level structure rather than one for child diagnostic as well, but this seemed easiest to implement given the current diagnostics framework. But I'm well aware that I've likely overlooked something.

**Comment by est31** [comment]
I've made PRs with proposed functionality for cargo as well as rustc. https://github.com/rust-lang/rust/pull/73945 https://github.com/rust-lang/cargo/pull/8437

`--extern-location` isn't needed by either of the PRs, and won't be even in the future when spans are printed as cargo synthesizes its own warnings already now. As the flag is useful for Buck, `--extern-location` should absolutely be added, but I suggest reducing it to what buck uses.

**Comment by nikomatsakis** [comment]
I've gone ahead and seconded the MCP.

**Comment by spastorino** [comment]
Given that this issue is now seconded, is it really waiting on team or we should just change to waiting on author/review?

**Comment by pnkfelix** [comment]
@spastorino I think that we still need to wait for the end of the final comment period for the MCP before we can say that this is "just" waiting on review.

I switched its state to waiting-on-fcp, in reference to the 10 day final comment period [here](https://github.com/rust-lang/compiler-team/issues/303)

**Comment by jsgf** [comment]
I edited the PR description to match the current state of this PR - that is, only providing `raw` and `json` location types, and default locations when `--extern-location` isn't provided.

**Comment by nikomatsakis** [comment]
So the MCP was accepted here long ago. I'm assigned to this PR, but I don't feel I'm the best reviewer. I'm *willing* to review, but @petrochenkov would you prefer to review?

**Comment by petrochenkov** [comment]
No, I'd prefer to leave this to someone else.

**Comment by JohnCSimon** [comment]
@jsgf  Ping from triage - this PR has some merge conflicts. Thanks.

**Comment by crlf0710** [comment]
@jsgf Ping from triage, there's still merge conflicts.

**Comment by jsgf** [comment]
@JohnCSimon @crlf0710 Looking.

**Review by nikomatsakis — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
@jsgf the CI is failing, can you resolve the issue ?

**Comment by jsgf** [comment]
@Dylan-DPC I'm trying to but the failure is in the Cargo tests in the cargo submodule. How can I sync that in a PR?

**Comment by pietroalbini** [comment]
@bors r-

The PR failed but bors forgot about that during synchronize.

**Comment by nikomatsakis** [comment]
@jsgf I guess [this is the failure](https://github.com/rust-lang-ci/rust/runs/1212658999#step:23:8920) that you're talking about? I can't quite tell what that's all about or if it's related to this PR exactly.

**Comment by nikomatsakis** [comment]
Maybe somebody from @rust-lang/cargo can help explain what's going on with the metabuild test.

**Comment by ehuss** [comment]
It looks like the test is rejecting the extra field `tool_metadata` in the JSON output. I would recommend not emitting that field if it is `null` (especially since this is an unstable feature).

It would also be nice to have some documentation for this feature in the unstable book.

**Comment by nikomatsakis** [comment]
Ah, yes, good catch re: the unstable book! @jsgf can you add an entry to https://github.com/rust-lang/rust/tree/master/src/doc/unstable-book for this feature as well?

**Comment by jsgf** [comment]
@ehuss Good point on unstable book. I'll update that.

I tried to not emit the `tool_metadata` field if its null, but that doesn't seem to be something the current serializer framework supports. I tried adding it, but it started getting very messy. Am I missing something? Are there other examples of null/None fields which are skipped?

File: compiler/rustc_errors/src/diagnostic.rs
Comment (bjorn3) [inline compiler/rustc_errors/src/diagnostic.rs:512]:
```suggestion
    /// Adds a suggestion intended only for a tool. The intent is that the metadata encodes
```

Comment (jsgf) [reply compiler/rustc_errors/src/diagnostic.rs:512]: <pedantry>
Technically "data" and hence "metadata" are plural, so "encode" is strictly correct.
</pedantry>
 But in common use we use data as a (singular) mass noun so "encodes" would be the right form to agree.

**Comment by nikomatsakis** [comment]
@jsgf 

> Are there other examples of null/None fields which are skipped?

Ugh. I admit that I haven't looked at this code in a (rather long...) while though, I presume we're using some kind of home-grown JSON encoding..? I can try to investigate a bit, it's quite possible that this isn't something we've supported.

**Comment by Dylan-DPC** [comment]
@jsgf any updates?

**Comment by jsgf** [comment]
@Dylan-DPC No, I'm still stalled on how to update the cargo test and sync it with the rust repo. I agree the best path would be to no emit null fields in the json which would clean up a lot of the output, but there doesn't seem to be a way to do it with the current json serialization framework. One option might be to port the serialization to use serde, which does have a way to do that, but that seems like a lot of work to resolve this particular problem.

**Comment by crlf0710** [comment]
@jsgf Ping from triage! Any updates on this?

File: compiler/rustc_errors/src/json.rs
Comment (nikomatsakis) [inline compiler/rustc_errors/src/json.rs:186]: OK @jsgf I took a look at this struct. I think the way to skip emitting the `tool_metadata` field would be to implement the `Encodable` trait *manually* here. It's kind of a pain, for sure, though I guess not *so* bad. You would basically write some code like

```rust
impl<E: Encoder> Encodable<E> for Diagnostic {
    fn encode(&self, s: &mut E) -> Result<(), S::Error> {
        s.emit_struct_field("message", 0, |s| self.message.encode(s))?;
        ...
        if self.tool_metadata.is_some() {
            s.emit_struct_field("message", N, |s| self.message.encode(s))?;
        }
        Ok(())
    }
}
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Sorry for being so slow @jsgf, care to take a stab at changes like this?

File: compiler/rustc_errors/src/json.rs
Comment (nikomatsakis) [reply compiler/rustc_errors/src/json.rs:186]: Obviously a macro could reduce some of the boilerplate here :)

**Comment by JohnCSimon** [comment]
Ping from triage: 
@jsgf - can you please address the merge conflicts and comments from nikomatsakis ? thank you !

File: compiler/rustc_errors/src/json.rs
Comment (jsgf) [reply compiler/rustc_errors/src/json.rs:186]: I did consider that, but I was concerned that it would become a maintenance liability (add a field without updating the encoder).

**Comment by jsgf** [comment]
OK I've updated to skip null `tool_metadata` entries which avoids changing a pile of unrelated test fixtures, and it looks like it's passing tests.

**Comment by jsgf** [comment]
@nikomatsakis ping?

**Comment by nikomatsakis** [comment]
@jsgf looking now! Sorry! :)

File: compiler/rustc_errors/src/json.rs
Comment (nikomatsakis) [reply compiler/rustc_errors/src/json.rs:188]: To avoid this, I usually use a desugaring like

```rust
let Struct { field1, field2 } = value;
execute(field1, field2);
```

this way, if fields are missing, you get a compilation error

File: compiler/rustc_errors/src/json.rs
Comment (nikomatsakis) [inline compiler/rustc_errors/src/json.rs:194]: Per [this point](https://github.com/rust-lang/rust/pull/72603/files#r559808807), maybe we can change the desugaring to

```rust
let $struct_name { $($name),+ } = $s;
let mut idx = 0;
$(
    $enc.emit_struct_field(
        stringify!($name),
        idx,
        |enc| $name.encode(enc),
    )?;
    idx += 1;
)+
idx
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great, I left one nit, apologies for the slow review @jsgf

**Comment by nikomatsakis** [comment]
@bors delegate+

I'm delegating to @jsgf to do `r=nikomatsakis` once the nit is addressed.

File: compiler/rustc_errors/src/json.rs
Comment (jsgf) [reply compiler/rustc_errors/src/json.rs:194]: Hm, that works, but since the whole point is that we're leaving out some fields to special-case them it will also need a list of those so they can be matched.

**Pull request merged**
